cftime/src/parser.rs

```diff
@@ -0,0 +1,391 @@
+            tag("hr"),
+            tag("h")
+        )),
+    );
+    #[rustfmt::skip]
+    let minutes = value(
+        CFDuration::minutes(1,*calendar),
+        alt((
+            tag("minutes"),
+            tag("minute"),
+            tag("mins"),
+            tag("min")
+        )),
+    );
+    let seconds = value(
+        CFDuration::seconds(1, *calendar),
+        alt((
+            tag("seconds"),
+            tag("second"),
+            tag("secs"),
+            tag("sec"),
+            tag("s"),
+        )),
+    );
+    let milliseconds = value(
+        CFDuration::milliseconds(1, *calendar),
+        alt((
+            tag("milliseconds"),
+            tag("millisecond"),
+            tag("millisecs"),
+            tag("millisec"),
+            tag("msecs"),
+            tag("msec"),
+            tag("ms"),
+        )),
+    );
+    let microseconds = value(
+        CFDuration::microseconds(1, *calendar),
+        alt((
+            tag("microseconds"),
+            tag("microsecond"),
+            tag("microsecs"),
+            tag("microsec"),
+        )),
+    );
+
+    alt((
+        years,
+        months,
+        days,
+        hours,
+        minutes,
+        seconds,
+        milliseconds,
+        microseconds,
+    ))(input)
+}
+
+/// macro created to not pass two argument to the function as it is a requirement
+/// for nom::separated_pair
+
+fn date<'a>(input: &'a str, calendar: &'a Calendars) -> IResult<&'a str, CFDates> {
+    let cf_date_factory = CFDateFactory {
+        calendar: *calendar,
+    };
+    let ymd = map(
+        tuple((i32, tag("-"), u32, tag("-"), u32)),
+        |(year, _, month, _, day)| cf_date_factory.build(year, month, day).unwrap(),
+    );
+
+    let x = alt((ymd,))(input);
+    x
+}
+fn time(input: &str) -> IResult<&str, Time> {
+    let hms = map(
+        tuple((i32, tag(":"), u32, tag(":"), double)),
+        |(hour, _, minute, _, second)| {
+            let (second, rest) = (second.trunc(), second.fract());
+            let nanosecond = rest * 1e9;
+
+            Time {
+                hour: hour as u32,
+                minute: minute as u32,
+                second: second as u32,
+                nanosecond: nanosecond as u64,
+            }
+        },
+    );
+
+    let hm = map(separated_pair(i32, tag(":"), u32), |(hour, minute)| Time {
+        hour: hour as u32,
+        minute: minute,
+        second: 0,
+        nanosecond: 0,
+    });
+
+    let x = alt((hms, hm))(input);
+    x
+}
+
+fn timezone(input: &str) -> IResult<&str, Tz> {
+    println!("{input}");
+    let hm = map(
+        preceded(opt(tag("+")), separated_pair(i8, tag(":"), u8)),
+        |(hour, minute)| Tz {
+            hour: hour,
+            minute: minute,
+        },
+    );
+    let z = value(Tz::default(), tag("Z"));
+    let utc = value(Tz::default(), tag("UTC"));
+    alt((hm, z, utc))(input)
+}
+
+fn datetime<'a>(input: &'a str, calendar: &'a Calendars) -> IResult<&'a str, CFDatetimes> {
```

Comment:
Take `Calendars` by value as this type is Copy

Reply:
The return type should be `NaiveDateTime` and the conversion can be done by mapping the calendar on the datetime in case the user wants to do time-calculations without the calendar

Reply:
The problem with NaiveDateTime from `chrono` is that it is based on the proleptic gregorian calendar, so date like `1600-02-30 00:00:00"` can not exist with this structure (but it can with 360 day calendar). NaiveDateTime throw an error.

I was thinking of `Option<Calendars>` to define a default calendar if None is supplied. What do you think ? 
Also, maybe instead of returning CFDatetime we could use generics to return struct that implement DateTimeLike trait

Reply:
We can do our own `NaiveDateTime` and not rely on `chrono`

Reply:
Sorry, i was thinking you were talking about `chrono`. But doing so, we will face the same problem, no ? We have at a moment or an other return something from `CFDatetimes` ? 
Also does it make sense to do calculcations on datetime without calendar ? I mean, how is it better than providing a default calendar ?

Reply:
We still have the problem of changing from datetime+calender+duration into a new datetime, but this is separate from parsing the datetime. I envision parsing into `NaiveDateTime`, interpreting as a calendar time (which might fail) then adding a duration to get a new calendar time

Reply:
Separation of concern, definitly makes sense